### DISCUSSION TITLE
How to make struct fields only modifiable by invoking their members?

### DISCUSSION THREAD

**Discussion by YT-GameMaker** [dotnet/runtime#96945] [Q&A]
```C#
class Random
{
    public static Random Shared { get; }
}
```
`System.Random` provides a shared instance where we can modify its state by invoking its members. I wrote a lightweight value-type `Random`, but found it challenging to provide a shared instance. If exposed directly, it could be assigned directly. If exposed through a get accessor, the instance's state cannot be modified. Is there a way to allow access to members of a struct field but disallow direct modification of the field itself?

**Comment by colejohnson66** [comment]


**Reply by YT-GameMaker** [reply]
This doesn't work, the `ref return` allows direct modification of the backing field, so you can do it like this: `MyRandom.Shared = default`.

**Reply by colejohnson66** [reply]
Ah yes. I forgot about that.

**Comment by Clockwork-Muse** [comment]
As a side note, shared static members (especially intentionally mutable ones) are nearly impossible to test with, and if you're using any sort of multithreading/concurrency (any sort of UI, `async`/`await`, etc) you can corrupt them pretty quickly.

Injecting it as a dependency at least makes it easier to test, although it doesn't absolve you of any concurrency concerns...

**Comment by YT-GameMaker** [comment]
I guess it's not achievable right now, perhaps a new syntax might be needed.

**Comment by En3Tho** [comment]
But what is the problem of using a class? Does you type has to be a struct no matter what? 

In you example you basically solve the problem by using a class. Shared instance can be modified (because it's a class) but the backing field of "Shared" itself can't.

From what I've understood you want "I want this variable to be not directly assignable/readonly but it's contents should be mutable nonetheless". This is what classes can give you. Or am I missing something here?

It is harder to do with a struct, something like a wrapper comes to mind. You can then combine ref on an original static field containing a struct and a another ref struct wrapper than acts as a proxy to provide an access to fields/members through refs too but hiding an original field. It is kinda clumsy though.

**Comment by DaZombieKiller** [comment]
I think this belongs on the [dotnet/csharplang](https://github.com/dotnet/csharplang/) repo. What you're asking for sounds like controlled-mutability references, which are currently used for `using` variables and `foreach` enumeration variables (you can call mutating members on them, but not re-assign the value entirely). There isn't currently a way to define your own controlled-mutability reference in C#.

An API I can think of which could have used such a feature in the past is `Unsafe.Unbox`: it returns a controlled-mutability reference, but today it can only be expressed as either `ref` (not restrictive enough) or `ref readonly` (too restrictive).
